### ISSUE TITLE
Feature request: upload source not local, but from URL

### ISSUE THREAD

**Issue by edoderoo** [yarl/pattypan#6]
In case files are stored online, it could be helpful if the Excel sheet can point out to a list of URLs instead of local file names.

**Comment by yarl** [comment]
Yes, this is definitely planned, should be ready for next release.

**Comment by Lanbot84** [comment]
Hi, I was wondering what the status of this request is? I am working at an archive for whom this would be very useful and would like to give them a time frame for the development if possible? Thanks so much!

**Comment by Abbe98** [comment]
I'm working on this actually, hoping to make a pull request this weekend.

**Comment by yarl** [comment]
Sweet, thanks @Abbe98. Just to be sure, you want to implement it using https://www.mediawiki.org/wiki/API:Upload#Uploading_from_URL?

**Comment by Abbe98** [comment]
Yep, it has been added go the .org.wikipedia package so I update it...

**Comment by Lanbot84** [comment]
Great @Abbe98, I am looking forward to it! Thanks so much!

**Comment by yarl** [comment]
Code is almost done, but it looks it won't work, due to Commons configuration. Check this ticket on Phabricator.

https://phabricator.wikimedia.org/T140040

**Comment by Reniervdg** [comment]
@yarl Thanks fot your work on this issue. Created an issue on Phabricator: https://phabricator.wikimedia.org/T181713 My domain got whitelisted, but uploading doesn't work yet, see my comment of Fri, Dec 8, 10:26 AM.

**Comment by Abbe98** [comment]
@Reniervdg are you using a custom build of Pattypan? Uploading by URL has not landed in a release yet.

**Comment by Reniervdg** [comment]
No, I'm not using a custom build. I was under the impression that uploading by URL was already integrated, but apparently I was wrong, my apologies. When is a new release scheduled, if I may ask?

**Comment by yarl** [comment]
@Reniervdg I will do my best to release is this week.

**Comment by Reniervdg** [comment]
@yarl Great, I think this will benefit a lot of people. 

I have tested Pattypan for my own project. However, when using it I receive this error: **"filename does not include a valid file extension"** The output from our archival information system comes in the form of this type of URL: https://studiezaal.nijmegen.nl/HttpHandler/icoon.ico?file=14102767 (as explained in the Phabricator-issue https://phabricator.wikimedia.org/T181713). Pattypan apparently cannot process this type of URL currently. Can this be remedied somehow?

**Comment by Abbe98** [comment]
@Reniervdg note that the `name` column in the spreadsheet needs to include and end with the file extension(in this case `.tif`).

I should write documentation for this feature... :-)

**Comment by Reniervdg** [comment]
@Abbe98 Thanks for the explanation. Validation of the spreadsheet is now succesful! However, when I go to actually upload a file + metadata, I receive the error **"Permission denied: cannot upload files via URL"** What's going on? I'm using Pattypan version 17.12.

**Comment by Abbe98** [comment]
@Reniervdg you need the user right `upload_by_url`, see:

https://commons.wikimedia.org/wiki/Commons:Upload_tools#Upload_by_URL

**Comment by Reniervdg** [comment]
I didn't know, thank you! I've applied for this user right via https://commons.wikimedia.org/wiki/Commons:Requests_for_rights#Reniervdg

**Comment by Reniervdg** [comment]
User right granted. :) Everything is working like a charm now. Thank you @yarl for adding the upload via URL-feature and thank you @Abbe98 for helping me along!

**Comment by Abbe98** [comment]
Reopened as it's still not documented.
